extensions/amp-cxense-player/0.1/amp-cxense-player.js

```diff
@@ -0,0 +1,319 @@
+      setStyles(this.iframe_, {display: 'none'});
+
+      this.element.appendChild(this.iframe_);
+      this.listenForPostMessages_();
+
+      return loadPromise(this.iframe_).then(ret => {
+        if (self.placeholder_) {
+          setStyles(this.placeholder_, {display: 'none'});
+        }
+        setStyles(self.iframe_, {display: ''});
+        self.applyFillContent(self.iframe_);
+        return ret;
+      });
+    }
+
+    /** @override */
+    pauseCallback() {
+      this.postMessage_({type: 'mpf.video.pause'});
+      return true;
+    }
+
+    /** @override */
+    unlayoutCallback() {
+      this.iframe_.setAttribute('src', 'about:blank');
+      this.iframe_.parentNode.removeChild(this.iframe_);
+      return true;
+    }
+
+    /** @private */
+    listenForPostMessages_() {
+      const hostRegExp = new RegExp(
+            '^' + parseUrl(cxDefaults.embedHost).hostname
+        );
+
+      this.getWin().addEventListener('message', e => {
+        if (e && e.data) {
+          let message = {data: {}};
+          try {
+            message = JSON.parse(e.data);
+          } catch (e) {}
+
+          if (!message.type
+                    || !message.data.location
+                    || !hostRegExp.test(message.data.location.hostname)
+                ) {
+            return;
+          }
+                // console.log("messge from iframe", message);
+                // todo: could implement a dispatch function here - not needed for now.
+                // this.dispatch_('')
+        }
+      }, false);
+    }
+
+    /** @private */
+    getIframeSrc_() {
+      const attrs = this.getExpandedDataAttributes_();
+      return appendParamStringToUrl(
+            cxDefaults.embedHost
+            + (attrs.debug
+                    ? cxDefaults.debugEmbedApp
+                    : cxDefaults.distEmbedApp
+            ),
+            queryString(this.getCollapsedDataAttributes_(attrs))
+        );
+    }
+
+    /** @private */
+    getExpandedDataAttributes_(element) {
+      element = element || this.element;
+      return extend(true,
+            {},
+            cxDefaults.attrs,
+            expandKeys(attrHash(element, 'data'))
+        );
+    }
+
+    /** @private */
+    getCollapsedDataAttributes_(hash) {
+      hash = hash || this.getExpandedDataAttributes_();
+      const newHash = {};
+      recurseObject(hash, function(keypath, value) {
+        newHash[keypath] = value;
+      });
+      return newHash;
+    }
+
+    /** @private */
+    postMessage_(data) {
+      data = extend(true, {
+        location: location,
+        width: this.getWin().offsetWidth,
+        height: this.getWin().offsetHeight,
+      }, data || {});
+
+      return this.iframe_.contentWindow.postMessage(
+            JSON.stringify(data), cxDefaults.embedHost
+        );
+    }
+
+    /** @private */
+    buildWidgetPlaceholder_() {
+      const doc = this.getDoc_();
+
+      this.placeholder_ = doc.createElement('div');
```

Comment:
That's the only thing left on the list, we're working on standing an API for that, do you think we can schedule this for another PR? or is it a blocker?